File: backend/src/utils/java.ts
import fs from "node:fs";
import { exec } from "node:child_process";
import { logger } from "./logger.js";

export interface JavaVersion {
  major: number;
  minor: number;
  patch: number;
  fullVersion: string;
  vendor: string;
  runtimeVersion?: string;
}

export interface JavaCheckResult {
  exists: boolean;
  version?: JavaVersion;
  error?: string;
}

export function version_compare(v1: string, v2: string): number {
  const a = v1.split(".").map(Number);
  const b = v2.split(".").map(Number);
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const av = a[i] || 0;
    const bv = b[i] || 0;
    if (av !== bv) return av > bv ? 1 : -1;
  }
  return 0;
}

export async function checkJava(javaPath?: string): Promise<JavaCheckResult> {
  try {
    const javaCmd = javaPath || "java";
    const output = await new Promise<string>((resolve, reject) => {
      exec(`${javaCmd} -version`, (err, stdout, stderr) => {
        if (err) {
          logger.error("Java 检查失败", err);
          reject(new Error("Java not found"));
          return;
        }
        resolve(stderr);
      });
    });

    logger.debug(`Java version output: ${output}`);

    const versionRegex = /version "(\d+)(\.(\d+))?(\.(\d+))?/;
    const vendorRegex = /(Java\(TM\)|OpenJDK).*Runtime Environment.*by (.*)/;

    const versionMatch = output.match(versionRegex);
    const vendorMatch = output.match(vendorRegex);

    if (!versionMatch) {
      return {
        exists: true,
        error: "解析 Java 版本失败",
      };
    }

    const major = parseInt(versionMatch[1], 10);
    const minor = versionMatch[3] ? parseInt(versionMatch[3], 10) : 0;
    const patch = versionMatch[5] ? parseInt(versionMatch[5], 10) : 0;

    const versionInfo: JavaVersion = {
      major,
      minor,
      patch,
      fullVersion: versionMatch[0].replace("version ", ""),
      vendor: vendorMatch ? vendorMatch[2] : "Unknown",
    };

    logger.info(`检测到 Java: ${JSON.stringify(versionInfo)}`);

    return {
      exists: true,
      version: versionInfo,
    };
  } catch (error) {
    logger.error("Java 检查异常", error as Error);
    return {
      exists: false,
      error: (error as Error).message,
    };
  }
}

export async function detectJavaPaths(): Promise<string[]> {
  const javaPaths: string[] = [];

  const windowsPaths = [
    "C:\\Program Files\\Java\\",
    "C:\\Program Files (x86)\\Java\\",
    "C:\\Program Files\\Eclipse Adoptium\\",
    "C:\\Program Files\\Eclipse Foundation\\",
    "C:\\Program Files\\Microsoft\\",
    "C:\\Program Files\\Amazon Corretto\\",
    "C:\\Program Files\\BellSoft\\",
    "C:\\Program Files\\Zulu\\",
    "C:\\Program Files\\Semeru\\",
    "C:\\Program Files\\Oracle\\",
    "C:\\Program Files\\RedHat\\",
  ];

  for (const basePath of windowsPaths) {
    try {
      if (fs.existsSync(basePath)) {
        const versions = fs.readdirSync(basePath);
        for (const version of versions) {
          const javaExe = `${basePath}${version}\\bin\\java.exe`;
          if (fs.existsSync(javaExe)) {
            javaPaths.push(javaExe);
          }
        }
      }
    } catch {
      // Directory might not be readable
    }
  }

  try {
    const pathOutput = await new Promise<string>((resolve, reject) => {
      exec("where java", (err, stdout) => {
        if (err) { resolve(""); return; }
        resolve(stdout);
      });
    });

    for (const p of pathOutput.split('\n').filter(p => p.trim())) {
      if (!javaPaths.includes(p.trim())) {
        javaPaths.push(p.trim());
      }
    }
  } catch {
    // `where` not available
  }

  return [...new Set(javaPaths)];
}

File: core/java/java.go
package java

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"deearthx/core/util"
)

// JavaVersion represents parsed Java version information
type JavaVersion struct {
	Major          int    `json:"major"`
	Minor          int    `json:"minor"`
	Patch          int    `json:"patch"`
	FullVersion    string `json:"fullVersion"`
	Vendor         string `json:"vendor"`
	RuntimeVersion string `json:"runtimeVersion,omitempty"`
}

// JavaCheckResult represents the result of a Java check
type JavaCheckResult struct {
	Exists  bool         `json:"exists"`
	Version *JavaVersion `json:"version,omitempty"`
	Error   string       `json:"error,omitempty"`
}

// VersionCompare compares two version strings
// Returns: -1 if v1 < v2, 0 if v1 == v2, 1 if v1 > v2
func VersionCompare(v1, v2 string) int {
	a := strings.Split(v1, ".")
	b := strings.Split(v2, ".")
	maxLen := max(len(a), len(b))

	for i := 0; i < maxLen; i++ {
		av := 0
		bv := 0
		if i < len(a) {
			av, _ = strconv.Atoi(a[i])
		}
		if i < len(b) {
			bv, _ = strconv.Atoi(b[i])
		}
		if av != bv {
			if av > bv {
				return 1
			}
			return -1
		}
	}
	return 0
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// CheckJava checks if Java is available and returns version info
func CheckJava(javaPath string) JavaCheckResult {
	cmd := javaPath
	if cmd == "" {
		cmd = "java"
	}

	// Run java -version
	execCmd := exec.Command(cmd, "-version")
	output, err := execCmd.CombinedOutput()
	if err != nil {
		util.Logger.Error(fmt.Sprintf("Java check failed: %v", err))
		return JavaCheckResult{
			Exists: false,
			Error:  err.Error(),
		}
	}

	util.Logger.Debug(fmt.Sprintf("Java version output: %s", output))

	// Parse version
	versionRegex := regexp.MustCompile(`version "(\d+)(\.(\d+))?(\.(\d+))?`)
	vendorRegex := regexp.MustCompile(`(Java\(TM\)|OpenJDK).*Runtime Environment.*by (.*)`)

	versionMatch := versionRegex.FindStringSubmatch(string(output))
	vendorMatch := vendorRegex.FindStringSubmatch(string(output))

	if versionMatch == nil {
		return JavaCheckResult{
			Exists: true,
			Error:  "Failed to parse Java version",
		}
	}

	major, _ := strconv.Atoi(versionMatch[1])
	minor := 0
	patch := 0
	if len(versionMatch) > 3 && versionMatch[3] != "" {
		minor, _ = strconv.Atoi(versionMatch[3])
	}
	if len(versionMatch) > 5 && versionMatch[5] != "" {
		patch, _ = strconv.Atoi(versionMatch[5])
	}

	vendor := "Unknown"
	if vendorMatch != nil && len(vendorMatch) > 2 {
		vendor = vendorMatch[2]
	}

	versionInfo := &JavaVersion{
		Major:          major,
		Minor:          minor,
		Patch:          patch,
		FullVersion:    strings.TrimPrefix(versionMatch[0], "version "),
		Vendor:         vendor,
		RuntimeVersion: string(output),
	}

	util.Logger.Info(fmt.Sprintf("Detected Java: %+v", versionInfo))

	return JavaCheckResult{
		Exists:  true,
		Version: versionInfo,
	}
}

// DetectJavaPaths scans common Java installation directories
func DetectJavaPaths() []string {
	var javaPaths []string

	// Common Windows Java installation paths
	windowsPaths := []string{
		`C:\Program Files\Java\`,
		`C:\Program Files (x86)\Java\`,
		`C:\Program Files\Eclipse Adoptium\`,
		`C:\Program Files\Eclipse Foundation\`,
		`C:\Program Files\Microsoft\`,
		`C:\Program Files\Amazon Corretto\`,
		`C:\Program Files\BellSoft\`,
		`C:\Program Files\Zulu\`,
		`C:\Program Files\Semeru\`,
		`C:\Program Files\Oracle\`,
		`C:\Program Files\RedHat\`,
	}

	for _, basePath := range windowsPaths {
		entries, err := os.ReadDir(basePath)
		if err != nil {
			continue
		}

		for _, entry := range entries {
			javaExe := filepath.Join(basePath, entry.Name(), "bin", "java.exe")
			if _, err := os.Stat(javaExe); err == nil {
				javaPaths = append(javaPaths, javaExe)
			}
		}
	}

	// Check PATH environment
	execCmd := exec.Command("where", "java")
	output, err := execCmd.Output()
	if err == nil {
		scanner := bufio.NewScanner(bytes.NewReader(output))
		for scanner.Scan() {
			path := strings.TrimSpace(scanner.Text())
			if path != "" && !contains(javaPaths, path) {
				javaPaths = append(javaPaths, path)
			}
		}
	}

	// Remove duplicates
	return unique(javaPaths)
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func unique(slice []string) []string {
	keys := make(map[string]bool)
	result := []string{}
	for _, item := range slice {
		if !keys[item] {
			keys[item] = true
			result = append(result, item)
		}
	}
	return result
}